consensus/src/runway/backup.rs

```diff
@@ -167,45 +172,114 @@ pub async fn run_loading_mechanism<'a, H: Hasher, D: Data, S: Signature, R: Read
         return;
     }
 
     let next_round_collection = match next_round_collection_rx.await {
         Ok(round) => round,
         Err(e) => {
-            error!(target: "AlephBFT-unit-backup", "unable to receive response from unit collections: {}", e);
+            error!(target: "AlephBFT-unit-backup", "Unable to receive response from unit collection: {}", e);
             on_shutdown(starting_round_tx);
             return;
         }
     };
-    info!(target: "AlephBFT-unit-backup", "received next round from unit collection: {:?}", next_round_collection);
+
+    info!(target: "AlephBFT-unit-backup", "Next round inferred from collection: {:?}", next_round_collection);
 
     if next_round_backup < next_round_collection {
-        error!(target: "AlephBFT-unit-backup", "backup lower than unit collection result. Backup got: {:?}, collection got: {:?}", next_round_backup, next_round_collection);
+        // Our newest unit doesn't appear in the backup. This indicates a serious issue, for example
+        // a different node running with the same pair of keys. It's safer not to continue.
+        error!(
+            target: "AlephBFT-unit-backup", "Backup state behind unit collection state. Next round inferred from: collection: {:?}, backup: {:?}",
+            next_round_collection,
+            next_round_backup,
+        );
         on_shutdown(starting_round_tx);
         return;
     };
 
     if next_round_collection < next_round_backup {
-        warn!(target: "AlephBFT-unit-backup", "unit collection result lower than backup. Backup got: {:?}, collection got: {:?}", next_round_backup, next_round_collection);
+        // Our newest unit didn't reach any peer, but it resides in our backup. One possible reason is that
+        // our node was taken down after saving the unit, but before broadcasting it. It is safe to continue.
+        warn!(
+            target: "AlephBFT-unit-backup", "Backup state ahead of than unit collection state. Next round inferred from: collection: {:?}, backup: {:?}",
+            next_round_backup,
+            next_round_collection
+        );
     }
 
     if let Err(e) = starting_round_tx.send(Some(next_round_backup)) {
-        error!(target: "AlephBFT-unit-backup", "could not send starting round: {:?}", e);
+        error!(target: "AlephBFT-unit-backup", "Could not send starting round: {:?}", e);
+    }
+}
+
+/// A task responsible for saving units into backup.
+/// It waits for units to appear in `units_to_save_rx`, and writes them to backup.
+/// It announces a successful write through `unit_saved_tx`.
+pub async fn run_saving_mechanism<'a, H: Hasher, D: Data, S: Signature, W: Write>(
+    mut unit_saver: UnitSaver<W, H, D, S>,
+    mut save_unit_rx: mpsc::UnboundedReceiver<UncheckedSignedUnit<H, D, S>>,
+    unit_saved_tx: mpsc::UnboundedSender<UncheckedSignedUnit<H, D, S>>,
+    mut terminator: Terminator,
+) {
+    let mut terminator_exit = false;
+    loop {
```

Comment:
There is one important difference between this service based implementation and previous version, i.e. in case of an error, this version will stop any further attempts at storing new units and exit, but previous was still trying to store them. Was this intentional? Anyway, both current and previous versions were early failing if `load_backup` found a single broken unit.

Reply:
I believe the **whole** application will exit, when we get an error in backup saver... which I didn't even think of to be honest. When saver has an error, it exits, and [here](https://github.com/Cardinal-Cryptography/AlephBFT/blob/A0-1720-full-unit-backup/consensus/src/runway/mod.rs#L1061) we break the `select` loop, and terminate all tasks below.

I believe that with current implementation it is the only correct thing to do, as backup saving sits between receiving `AddedToDag`, and broadcasting units to the network. When backup saving crashes, we won't send units to the network anymore.

But we should definitely think about changing that, i.e. we could implement this in a way so that:
1. We don't exit the whole app when backup exits
2. We send units regardless of the state of backup saver (working / exited)

Regarding the problem you stated, we should consider it. I can imagine a scenario in which saving something failed, but the save was repeated or something so in the end our backup will be correct.

Reply:
Great observation btw

Reply:
I think having it exit in this situation is fine, nothing in the backup should fail unless we are in pretty dire circumstances (broken disk, completely running out of RAM, etc.). We could alternatively print a warning and keep running without saving at all, but I don't think this has that many advantages.

Reply:
I'm gonna leave it as it is.